pkg/images/enricher/enricher.go

```diff
@@ -120,7 +120,7 @@ type ImageEnricher interface {
 	EnrichImage(ctx context.Context, enrichCtx EnrichmentContext, image *storage.Image) (EnrichmentResult, error)
 	// EnrichWithVulnerabilities will enrich an image with its components and their associated vulnerabilities only.
 	// This will always force re-enrichment and not take existing values into account.
-	EnrichWithVulnerabilities(image *storage.Image, components *scannerV1.Components, notes []scannerV1.Note) (EnrichmentResult, error)
+	EnrichWithVulnerabilities(image *storage.Image, components any, notes []scannerV1.Note) (EnrichmentResult, error)
```

Comment:
This is a bit unfortunate, how about creating a wrapper struct type which has both component types (v1 & v4) as fields and exposes a method of `GetComponents()` which returns either v1 or v4, based on which field is non-nil? This way we at least avoid the `any` in the interface and make it a bit more clear what we expect (given also that _eventually_ this should get back to V1 components in... however many moons).

Reply:
Considered similar logic in the design, we added the `indexer_version` because nil components does not mean one scanner vs the other in all cases.  

If I'm understanding correctly the proposal - move the logic from the newly added `enrichWithVulnerabilities` method of the image service (which chooses components based on `indexer_version`) into this new `GetComponents()` method (which would still return `any` right?) - this new struct would be passed to the scanners via `GetVulnerabilities()` - and each scanner would still check the type of the components before using them.  We would need a way to identify in the return of `GetVulnerabilities()` if the components could be handled (or add to the new components struct a `Type()` method and only send components to a scanner if we know that scanner can handle them, but I'd rather that decision come from the individual scanner vs. outside the scanner so that the logic is encapsulated)

I had initially added a `bool` return to `GetVulnerabilities()` to indicate if the scanner could handle the components so that the loop in the enricher could proceed to the next scanner without erroring out. However I felt that was 'ugly' and opted for the `CanHandle` method which made the logic and returns cleaner (IMO).

Did I understand the proposal correctly? What are your thoughts?

Reply:
Yeah my point was rather along the lines of:

Given we have struct such as:
```go
type ScanComponents struct {
  v1 *v1.Components
  v4 *v4.Components
}
```

Instead of passing `components any` to `GetVulnerabilities`, we pass `components ScanComponents` to the scanner and within `GetVulnerabilities` we can have some helper like `Type()` exposed on the `ScanComponents` struct and decide _within_ `GetVulnerabilities` whether this can be scanned / is compatible with the given scanner type or not.

Currently this happens within the enricher by calling `vulnScanner.CanHandle(components)`, but I'd say this should be embedded within `GetVulnerabilities` (one can still offer clients like the enricher to react on "component mismatch" by returning a specific error type, but the logic of "can scanner X handle component type Y" should be done in the single place of the scanner type implementation IMO.

Reply:
> Currently this happens within the enricher by calling vulnScanner.CanHandle(components), but I'd say this should be embedded within GetVulnerabilities (one can still offer clients like the enricher to react on "component mismatch" by returning a specific error type, but the logic of "can scanner X handle component type Y" should be done in the single place of the scanner type implementation IMO.

Can't say I like the style of using errors for flow control when there is no actual error.  I took a slightly different approach of having `ScanComponents` knowledgeable about the scanner type that the components are for which avoids having to type cast.  Personally I still prefer to original `CanHandle()` as that I believe is cleaner.  Please take a look and let me know if that's acceptable, or if you would prefer the error approach.